fe/fe-core/src/main/java/com/starrocks/summary/QueryHistoryMgr.java

```diff
@@ -0,0 +1,142 @@
+import com.starrocks.qe.SimpleExecutor;
+import com.starrocks.sql.plan.ExecPlan;
+import com.starrocks.statistic.StatisticUtils;
+import com.starrocks.statistic.StatsConstants;
+import com.starrocks.thrift.TResultSinkType;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpStatus;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+public class QueryHistoryMgr {
+    private static final Logger LOG = LogManager.getLogger(QueryHistoryMgr.class);
+
+    private static final int CONVERT_BATCH_SIZE = 100;
+
+    private static final int MAX_MEMORY_SIZE = 200 * 1024 * 1024; // 200 MB
+
+    private static final ExecutorService EXECUTOR = ThreadPoolManager.newDaemonFixedThreadPool(
+            1, Integer.MAX_VALUE,
+            "query-history-executor", true);
+
+    private final List<QueryHistory> histories = Lists.newArrayListWithCapacity(CONVERT_BATCH_SIZE + 1);
+
+    private StringBuffer buffer = new StringBuffer("[");
+
+    private int batchSize = 0;
+
+    private LocalDateTime lastLoadTime = LocalDateTime.now();
+
+    public LocalDateTime getLastLoadTime() {
+        return lastLoadTime;
+    }
+
+    public synchronized void addQueryHistory(ConnectContext ctx, ExecPlan plan) {
+        if (!GlobalVariable.isEnableQueryHistory()) {
+            return;
+        }
+        if (ctx.isStatisticsConnection() || ctx.isStatisticsJob() || plan.getScanNodes().isEmpty()
+                || StringUtils.containsIgnoreCase(ctx.getExecutor().getOriginStmtInString(),
+                StatsConstants.INFORMATION_SCHEMA)) {
+            return;
+        }
+
+        histories.add(new QueryHistory(ctx, plan));
+        if (histories.size() > CONVERT_BATCH_SIZE || lastLoadTime.plusSeconds(
+                GlobalVariable.queryHistoryLoadIntervalSeconds).isBefore(LocalDateTime.now())) {
+            loadQueryHistory(Lists.newArrayList(histories));
+            histories.clear();
+        }
+    }
+
+    private void loadQueryHistory(List<QueryHistory> list) {
+        if (!StatisticUtils.checkStatisticTables(List.of(StatsConstants.QUERY_HISTORY_TABLE_NAME))) {
+            return;
+        }
+
+        ConnectContext ctx = StatisticUtils.buildConnectContext();
+        EXECUTOR.submit(() -> {
+            ctx.setThreadLocalInfo();
+            for (QueryHistory query : list) {
+                try {
+                    buffer.append(query.toJSON()).append(",");
+                } catch (Exception e) {
+                    LOG.warn("Failed to convert query history {}", query.getOriginSQL(), e);
+                }
+            }
+            try {
+                batchSize += list.size();
+
+                if (buffer.isEmpty()) {
+                    return;
+                }
+                if (buffer.length() < MAX_MEMORY_SIZE && lastLoadTime.plusSeconds(
```

Comment:
> ?

Reply:
is `less`, not `greate`, will load when exceed the memory or reach the time